exercises/t_jubb/ex02_oo_basics/src/main.cpp

```diff
@@ -0,0 +1,270 @@
+
+#include <algorithm>
+#include <cmath>
+#include <iostream>
+#include <string>
+#include <vector>
+
+
+/*
+Shape class
+ */
+class Shape {
+
+public:
+	Shape() { area = 0.0; this->perimeter = 0.0; this->sides = 0; this->type = " "; };
+
+	double get_area() { return area; };
+	double get_perimeter() { return perimeter; };
+	double get_sides() { return sides; };
+	std::string get_type() { return type; };
+	void print_shape() {  
+		int pad = 10 - type.size();
+		std::string pad_blank(pad, ' ');
+		std::cout << "Shape : " << type << pad_blank;
+		std::cout << " Area : " << area << " Perim. : " << perimeter << std::endl;
+		};
+
+protected:
+	double area, perimeter;
+	int sides;
+	std::string type;
+};
+
+/*
+Square class
+*/
+class Square: public Shape {
+
+public:
+	Square() {
+		height = 0.0;
+		sides = 4;
+		area = 0.0;
+		perimeter = 0.0;
+		type = "Square";
+	};
+	Square(double height) {
+		type = "Square";
+		sides = 4;
+		area = height * height;
+		perimeter = 4 * height;
+	};
+
+private:
+	double height;
+};
+
+/*
+Rectangle class
+*/
+class Rectangle : public Shape {
+
+public:
+	Rectangle() {
+		type = "Rectangle";
+		height = 0.0;
+		width = 0.0;
+		sides = 4;
+		area = 0.0;
+		perimeter = 0.0;
+	};
+	Rectangle(double height, double width) {
+		type = "Rectangle";
+		sides = 4;
+		area = height * width;
+		perimeter = 2 * height + 2 * width;
+	};
+
+private:
+	double height, width;
+
+};
+
+/*
+Triangle Class
+*/
+class IsoscelesTriangle : public Shape {
+
+public:
+	IsoscelesTriangle() {
+		type = "Triangle";
+		height = 0.0;
+		base = 0.0;
+		sides = 3;
+		area = 0.0;
+		perimeter = 0.0;
+	};
+	IsoscelesTriangle(double base, double height) {
+		type = "Triangle";
+		sides = 3;
+		area = 0.5 * base * height;
+		perimeter = base + 2 * sqrt(pow(height, 2) + 0.25 * pow(base, 2));
+	};
+
+private:
+	double base, height;
+};
+
+/*
+Circlee Class
+*/
+class Circle : public Shape {
+
+public:
+	Circle() {
+		type = "Circle";
+		radius = 0.0;
+		sides = 1;
+		area = 0.0;
+		perimeter = 0.0;
+	};
+	Circle(double radius) {
+		type = "Circle";
+		sides = 1;
+		area = M_PI * pow(radius, 2);
+		perimeter = M_PI * 2 * radius;
+	};
+
+private:
+	double radius;
+};
+
+/*
+Comparisons between shapes for area, perimeter
+*/
+bool compare_area(Shape& shape1, Shape& shape2) {
+	return shape1.get_area() < shape2.get_area();
+};
+bool compare_perimeter(Shape& shape1, Shape& shape2) {
+	return shape1.get_perimeter() < shape2.get_perimeter();
+};
+
+class ShapeSorter : public Shape {
```

Comment:
Is `ShapeSorter` a `Shape`. Public inheritance should be thought of as an "is-a" relationship, i.e a circle is a shape in the case above but here a `ShapeSorter` is not a shape.